src/traces/splom/attributes.js

```diff
@@ -0,0 +1,121 @@
+/**
+* Copyright 2012-2018, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var scatterGlAttrs = require('../scattergl/attributes');
+var cartesianIdRegex = require('../../plots/cartesian/constants').idRegex;
+
+function makeAxesValObject(axLetter) {
+    return {
+        valType: 'info_array',
+        freeLength: true,
+        role: 'info',
+        editType: 'calc',
+        items: {
+            valType: 'enumerated',
+            values: [cartesianIdRegex[axLetter].toString(), ''],
+            editType: 'plot'
+        },
+        description: [
+            '..'
+        ].join(' ')
+    };
+}
+
+module.exports = {
+    dimensions: {
+        _isLinkedToArray: 'dimension',
+
+        visible: {
+            valType: 'boolean',
+            role: 'info',
+            dflt: true,
+            editType: 'calc',
+            description: ''
+        },
+        label: {
+            valType: 'string',
+            role: 'info',
+            editType: 'calc',
+            description: ''
+        },
+        values: {
+            valType: 'data_array',
+            role: 'info',
+            editType: 'calc+clearAxisTypes',
+            description: [
+                ''
+            ].join(' ')
+        },
+
+        // TODO should add an attribute to pin down x only vars and y only vars
+        // like https://seaborn.pydata.org/generated/seaborn.pairplot.html
+        // x_vars and y_vars
+
+        editType: 'calc+clearAxisTypes',
+        description: [
+            '..'
+        ].join(' ')
+    },
+
+    mode: scatterGlAttrs.mode,
+    text: scatterGlAttrs.text,
+
+    marker: scatterGlAttrs.marker,
+
+    line: scatterGlAttrs.line,
+    connectgaps: scatterGlAttrs.connectgaps,
+
+    xdirection: {
+        valType: 'enumerated',
+        values: ['right', 'left'],
+        dflt: 'right',
+        role: 'info',
+        editType: 'plot',
+        description: ''
+    },
+    ydirection: {
+        valType: 'enumerated',
+        values: ['top', 'bottom'],
+        dflt: 'bottom',
+        role: 'info',
+        editType: 'plot',
+        description: ''
+    },
+
+    xaxes: makeAxesValObject('x'),
+    yaxes: makeAxesValObject('y'),
+
+    showdiagonal: {
+        valType: 'boolean',
+        role: 'info',
+        dflt: true,
+        editType: 'plot',
+        description: ''
+    },
```

Comment:
Eventually we will want the diagonals to support various types: scattergl, histogram, box, violin. So I'm wondering if instead of `showdiagonal` we should have `diagonal: {visible: (true|false)} to support future expansion:
```js
diagonal: {
    visible: (true|false),
    type: (scattergl|histogram|box|violin),
    // then attributes specific to each type, as well as some overrides per dimension
}
```

Reply:
> Eventually we will want the diagonals to support various types: scattergl, histogram, box, violin.

Interesting. So you're thinking a single splom trace could generate scatter panels AND histogram or box or violin traces on the diagonal? That would put us on-par with some of the seaborn. Yeah nice. I can't really think of any drawbacks from that approach except this might cause headaches to the workspace team.

Reply:
> a single splom trace could generate scatter panels AND histogram or box or violin traces on the diagonal?

Yes exactly. An extra complication is that this would create even more x or y axes (for the histogram count or box/violin category), depending on the type and orientation of these diagonal plots; but once we do that, it seems like the implementation should be fairly straightforward. But anyway none of that has to happen now, as long as we set up the attributes we DO need now so they'll support this.

Reply:
hi @alexcjohnson, i'm really interested to use the diagonal type. Is it available to use?

Reply:
Currently the diagonal can only have scatter type. Follow #2555 for more info, but we do not have any immediate plans for when other types will be added.

Reply:
oh okay. Thanks